packages/react-devtools-extensions/src/contentScripts/prepareInjection.js

```diff
@@ -96,56 +104,27 @@ window.addEventListener('pageshow', function({target}) {
   chrome.runtime.sendMessage(lastDetectionResult);
 });
 
-const detectReact = `
-window.__REACT_DEVTOOLS_GLOBAL_HOOK__.on('renderer', function({reactBuildType}) {
-  window.postMessage({
-    source: 'react-devtools-detector',
-    reactBuildType,
-  }, '*');
-});
-`;
-const saveNativeValues = `
-window.__REACT_DEVTOOLS_GLOBAL_HOOK__.nativeObjectCreate = Object.create;
-window.__REACT_DEVTOOLS_GLOBAL_HOOK__.nativeMap = Map;
-window.__REACT_DEVTOOLS_GLOBAL_HOOK__.nativeWeakMap = WeakMap;
-window.__REACT_DEVTOOLS_GLOBAL_HOOK__.nativeSet = Set;
-`;
-
-// If we have just reloaded to profile, we need to inject the renderer interface before the app loads.
-if (sessionStorageGetItem(SESSION_STORAGE_RELOAD_AND_PROFILE_KEY) === 'true') {
-  const rendererURL = chrome.runtime.getURL('build/renderer.js');
-  let rendererCode;
-
-  // We need to inject in time to catch the initial mount.
-  // This means we need to synchronously read the renderer code itself,
-  // and synchronously inject it into the page.
-  // There are very few ways to actually do this.
-  // This seems to be the best approach.
-  const request = new XMLHttpRequest();
-  request.addEventListener('load', function() {
-    rendererCode = this.responseText;
-  });
-  request.open('GET', rendererURL, false);
-  request.send();
-  injectCode(rendererCode);
-}
+// The legacy way to inject the global hook in Manifest V2 extensions.
+// In V3, we use chrome.scripting.registerContentScripts instead (see background.js)
+if (IS_FIREFOX) {
+  // If we have just reloaded to profile, we need to inject the renderer interface before the app loads.
+  if (
+    sessionStorageGetItem(SESSION_STORAGE_RELOAD_AND_PROFILE_KEY) === 'true'
+  ) {
+    injectCodeSync(chrome.runtime.getURL('build/renderer.js'));
```

Comment:
@lunaruan originally we inject renderer code conditionally (only when that session storage value is `'true'`) like this. However, as we change to register it in background.js in V3, there's seem to be no way we can check this value. I've changed it to conditionally **run** it in renderer.js.
In theory this change would slow down the page when you don't need to load renderer.js, but in my test run the impact looks pretty minimal.